File: src/kotlin/node.kt
package com.algorithms.trees.bst;

/**
 * Describes a node that can be associated with
 * a binary search tree.
 */
data class Node<T>(val value: T) {
  var data             = value;
  var left: Node<T>?   = null;
  var right: Node<T>?  = null;
  var parent: Node<T>? = null;
}

File: src/typescript/lib/node.ts
import { ITree } from './tree';

/**
 * Describes a node that can be associated with
 * a binary search tree.
 */
export class Node<T> {
    public value: T;
    public left: Node<T> | null;
    public right: Node<T> | null;
    public parent: Node<T> | null;
    public tree: ITree<T> | null;

    /**
     * Node constructor.
     * @param value the value to associate with the node.
     */
    constructor(value: T) {
      this.value = value;
      this.left = null;
      this.right = null;
      this.parent = null;
      this.tree = null;
    }
}
